src/telemetry/transport/mod.rs

```diff
@@ -194,63 +252,141 @@ impl FmtMetrics for Report {
         }
 
         tcp_open_total.fmt_help(f)?;
-        tcp_open_total.fmt_scopes(f, metrics.iter(), |m| &m.open_total)?;
+        metrics.fmt_by(f, tcp_open_total, |m| &m.open_total)?;
 
         tcp_open_connections.fmt_help(f)?;
-        tcp_open_connections.fmt_scopes(f, metrics.iter(), |m| &m.open_connections)?;
+        metrics.fmt_by(f, tcp_open_connections, |m| &m.open_connections)?;
 
         tcp_read_bytes_total.fmt_help(f)?;
-        tcp_read_bytes_total.fmt_scopes(f, metrics.iter(), |m| &m.read_bytes_total)?;
+        metrics.fmt_by(f, tcp_read_bytes_total, |m| &m.read_bytes_total)?;
 
         tcp_write_bytes_total.fmt_help(f)?;
-        tcp_write_bytes_total.fmt_scopes(f, metrics.iter(), |m| &m.write_bytes_total)?;
+        metrics.fmt_by(f, tcp_write_bytes_total, |m| &m.write_bytes_total)?;
 
         tcp_close_total.fmt_help(f)?;
-        tcp_close_total.fmt_scopes(f, metrics.iter_eos(), |e| &e.close_total)?;
+        metrics.fmt_eos_by(f, tcp_close_total, |e| &e.close_total)?;
 
         tcp_connection_duration_ms.fmt_help(f)?;
-        tcp_connection_duration_ms.fmt_scopes(f, metrics.iter_eos(), |e| &e.connection_duration)?;
+        metrics.fmt_eos_by(f, tcp_connection_duration_ms, |e| &e.connection_duration)?;
 
         Ok(())
     }
 }
 
-// ===== impl Key =====
+// ===== impl Sensor =====
 
-impl FmtLabels for Key {
-    fn fmt_labels(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        ((self.proxy, self.peer), self.tls_status).fmt_labels(f)
+impl Sensor {
+
+    pub fn open(metrics: Option<Arc<Mutex<Metrics>>>) -> Self {
+        if let Some(ref m) = metrics {
+            if let Ok(mut m) = m.lock() {
+                m.open_total.incr();
+                m.open_connections.incr();
+            }
+        }
+        Self {
+            metrics,
+            opened_at: Instant::now(),
+        }
+    }
+
+    pub fn record_read(&mut self, sz: usize) {
+        if let Some(ref m) = self.metrics {
+            if let Ok(mut m) = m.lock() {
+                m.read_bytes_total += sz as u64;
+            }
+        }
+    }
+
+    pub fn record_write(&mut self, sz: usize) {
+        if let Some(ref m) = self.metrics {
+            if let Ok(mut m) = m.lock() {
```

Comment:
This seems like the place where we'd benefit the most from being able to increment counters without locking; it would sure be nice to not have to take a lock on every read/write.